[Keyboard] Add Pimoroni Keybow 2040 (#21481)
qmk/qmk_firmware: keyboards/pimoroni/keybow2040/readme.md

> Keybow 2040
> ===========

Reviewer: ```suggestion
# Keybow 2040
```

The [Pimoroni Keybow 2040][keybow2040] is a 16 key mechanical macropad
controlled by a Raspberry Pi RP2040, with Kailh hot swap sockets and
per key RGB lighting, in a decorative FR4 sandwich. The Keybow 2040 is
also known by its codename Venus Dust, or its product codes of the
form PIM56x.

* Keyboard Maintainer: [Tony Finch](https://github.com/fanf2)
* Hardware Supported: Keybow 2040
* Hardware Availability: [Pimoroni][keybow2040]

[keybow2040]: https://shop.pimoroni.com/products/keybow-2040

Make example for this keyboard (after setting up your build environment):

    make pimoroni/keybow2040:default

Flashing example for this keyboard:

    make pimoroni/keybow2040:default:flash

See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools)
and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide)
for more information. Brand new to QMK? Start with our
[Complete Newbs Guide](https://docs.qmk.fm/#/newbs).

> bootloader
> ----------
> 
> The procedure to flash the firmware is similar to other RP2040
> devices. You can enter the bootloader in two ways:
> 
>   * **Vulcan nerve pinch**: hold down BOOTSEL while pressing RESET
>     (not so great if you have configured BOOTSEL to do something)
> 
>   * **Mr Resetti double-tap**: press RESET twice quickly
>     (only works when running the QMK firmware)
> 
> Find the BOOTSEL and RESET buttons on your Keybow 2040 either side of
> the USB socket on the edge of the PCB. When the USB socket is facing
> away from you, BOOTSEL is on the right and RESET is on the left.
> According to the schematic the reset button is called Mr Resetti.
> 
>   * _Aside_: The BOOTSEL button disables the QSPI flash that stores
>     the firmware, which makes the RP2040 use its built-in UF2
>     bootloader.
> 
> A USB drive will appear called `RPI-RP2`. Copy the firmware file
> `pimoroni_keybow2040_default.uf2` onto the `RPI-RP2` drive using
> whatever software is convenient (e.g. the `cp` command, the macOS
> Finder, or Windows Explorer). The `make :flash` command will copy the
> firmware for you.
> 
> The `RPI-RP2` drive will disappear (which might make your OS complain
> that you failed to unmount it first) and a keyboard device will appear
> in its place.
> 
> The Pimoroni web site has [instructions for reinstalling CircuitPython
> on the Keybow 2040](https://learn.pimoroni.com/article/circuitpython-and-keybow-2040)
> 

Reviewer: ```suggestion
## Bootloader 

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Double tap the RESET button, or hold BOOTSEL and tap RESET

The Pimoroni web site has [instructions for reinstalling CircuitPython
on the Keybow 2040](https://learn.pimoroni.com/article/circuitpython-and-keybow-2040)
```

Author: This change is wrong, because I have not enabled the bootmagic feature:

  * there's no point since the reset and bootsel buttons are easily accessible on the side of the Keybow2040
  * the RP2040's ROM implements the bootsel feature, so there's no need for QMK to do so
  * the bootsel button is not (0,0) in the matrix because that would mess up the 4x4 layout

I've trimmed the tutorial parts of that section, and added a sentence about bootsel vs bootmagic.

I've left my jokes in because this is supposed to be fun 🤪

Reviewer: Bootmagic doesn't just reset the board and jump to bootloader, it also clears the eeprom.  Which can be very useful.  And has nothing to do with the bootsel button (it's just an arbitrary spot in the matrix, but defaults to 0,0).

> notes
> -----
> 
>   * The BOOTSEL button doubles as a 17th configurable key.
> 
>   * The default keymap has F13 - F20 on the top half of the board
>     (nearest the USB socket), and a navigation cluster (arrows, home,
>     end, page up/down) on the bottom half. The BOOTSEL button changes
>     the RGB matrix mode.
> 
>   * The Keybow 2040 CircuitPython firmware uses VID = 0x16D0, PID = 0x08C6,
>     but this QMK firmware has different USB IDs to avoid confusing drivers:
>     VID = 0x5069 ('Pi'), PID = 0x4784 (4 'x' 4)
> 

Reviewer: Not sure how useful this is here. 
```suggestion
```

Author: I've left that in because there ought to be somewhere for users to find out what the default keymap is without reading the JSON, and I need somewhere to explain the info.json configuration choices. JSON isn't supposed to have comments, so 🤷

> resources
> ---------
> 
> Here are some links to the information I used when configuring QMK for the Keybow 2040.
> 
>   * [Piomoroni Keybow 2040 product page](https://shop.pimoroni.com/products/keybow-2040)
>   * [Keybow 2040 assembly instructions](https://learn.pimoroni.com/article/assembling-keybow-2040)
> 
> Places to find information about the Keybow 2040 hardware:
> 
>   * [schematic](https://cdn.shopify.com/s/files/1/0174/1800/files/keybow_2040_schematic.pdf)
>       * contains most of the information that can also be gleaned from the CircuitPython support
> 
>   * Pimoroni user-friendly CircuitPython libraries
>       * [old](https://github.com/pimoroni/keybow2040-circuitpython) and [new](https://github.com/pimoroni/pmk-circuitpython)
>       * these are wrappers around the low-level support below
> 
>   * [CircuitPython hardware support](https://github.com/adafruit/circuitpython/blob/main/ports/raspberrypi/boards/pimoroni_keybow2040/)
>       * defines map from hardware feature names (e.g. `SW15`) to the way they are accessed (e.g. GPIO6)
> 
>   * [CircuitPython IS31FL3731 driver](https://github.com/adafruit/Adafruit_CircuitPython_IS31FL3731/blob/main/adafruit_is31fl3731/keybow2040.py)
>       * defines the RGB matrix layout

Reviewer: This would be better linked in the "hardware supported" section above. Though, it looks like most of this is linked in the product page, already. 
```suggestion
```

Author: I have left these links in because it was difficult to find the relevant parts of the default Python firmware, so I think it will be helpful when anyone else wants to work on this board.

Reviewer: But this isn't the circuit python repo, so has zero relevance here.  And most of the links add no value.

_This QMK configuration was created by Tony Finch <dot@dotat.at>, **not** by Pimoroni._

The [Pimoroni Keybow 2040][keybow2040] is a 16 key mechanical macropad
controlled by a Raspberry Pi RP2040, with Kailh hot swap sockets and
per key RGB lighting, in a decorative FR4 sandwich.

The Keybow 2040 is also known by its codename Venus Dust, or its
product codes of the form PIM56x.

[keybow2040]: https://shop.pimoroni.com/products/keybow-2040

Make example for this keyboard (after setting up your build environment):

    make pimoroni/keybow2040:default

Flashing example for this keyboard:

    make pimoroni/keybow2040:default:flash

See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools)
and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide)
for more information. Brand new to QMK? Start with our
[Complete Newbs Guide](https://docs.qmk.fm/#/newbs).

notes
-----

  * The BOOTSEL button doubles as a 17th configurable key.

  * There is no USB PID allocated for the Keybow 2040 under
    [the Raspberry Pi USB VID](https://github.com/raspberrypi/usb-pid),
    but its CircuitPython port uses VID = 0x16D0, PID = 0x08C6

>     so I copied that.

Reviewer 2: I'm not sure whether copying the USB ID from CircuitPython is the proper thing to do — this may create some conflicts, like VIA trying to poke a board that uses the CircuitPython firmware and therefore doing some unexpected things.

Author: Oh good, I thought it might be a bad idea too. (Also the CircuitPython VID is a real VID and the PID only fails to collide by luck.) I will go back to my original plan of cooking up a new “random” pair of IDs.

Reviewer 2: 0x16D0 is the VID for which unique PID values are sold by [MCS Electronics](https://www.mcselec.com/index.php?page=shop.product_details&product_id=92&option=com_phpshop), so the PID can probably be considered as “owned” by Pimoroni.

The problem with VIA is that it does not perform a proper detection for the HID usage page to find the rawhid interface (although that usage page is in the vendor range, so in theory it might collide with some completely unrelated usage).  Another potential problem is that Windows in some circumstances can cache the binding of drivers to USB interfaces, therefore you may have some trouble if the device changes the set of advertised interfaces without changing its USB VID:PID pair.

Author: Oh, I didn’t know MCS did that, cool. And thanks for those details about why it is best to avoid this kind of collision.

Author: I picked new USB IDs in round two, and I checked [it does not collide](https://devicehunt.com/search/type/pci/vendor/07F8/device/any)
